sched: Fix smp scheduling

jlaitine opened the pull request:
## Summary

This fixes scheduling issues found in both code review and testing. The main issues started with nxsched_addreadytorun putting tasks to the pending list on wrong basis; basically checking if the current CPU was executing in sched lock, when it was supposed to check that the selected CPU was executing a task in sched lock.

The other changes in this PR are related to the same issue, in different places of the code, checking scheduler locked on the wrong task/cpu.

On some of our platforms (arm64 smp 2 cores CONFIG_BUILD_KERNEL) this led to erroneous function, when sometimes at boot the NSH executing the boot script ended up into pending list, and no-one ever picked it back from there.

These issues have appeared already in https://github.com/apache/nuttx/commit/b4cc9fb11b7f51245f2f4a0c3ebb4a665205b320#diff-0ccba5c8654b46ae886ab122000258cad1ee91099ed0d0bf225731df6ff06426 , where the operation of the sched lock was changed.

## Impact

Impacts all SMP targets

## Testing

Tested on real application on custom hardwares on:
MPFS (risc-v 64) single core, CONFIG_BUILD_FLAT
IMX93 (arm64) single core, CONFIG_BUILD_FLAT
MPFS SMP 4 cores, CONFIG_BUILD_FLAT
MPFS SMP 4 cores, CONFIG_BUILD_KERNEL
IMX93 SMP 2 cores, CONFIG_BUILD_FLAT
IMX93 SMP 2 cores, CONFIG_BUILD_KERNEL

Tested on ostest on:
rv-virt:smp
rv-virt:smp64
modified rv-virt:smp64, "CONFIG_PRIORITY_INHERITANCE=y" added to config

Long period reliability testing is still ongoing on real hw.

jlaitine commented:
I realized that sched_mergepending didn't re-check the minimum priority after every loop. So added that...

hujun260 commented on sched/sched/sched_setpriority.c:
```diff
@@ -259,7 +250,7 @@ static inline void nxsched_running_setpriority(FAR struct tcb_s *tcb,
           if (nxsched_reprioritize_rtr(tcb, sched_priority))
             {
-              up_switch_context(this_task(), rtcb);
+              up_switch_context(this_task(), tcb);
```
why change rtcb ->tcb 
The second parameter of up_switch_context should be the running task which will be blocked.

jlaitine replied:
The tcb is the running task, which is having it's priority adjusted. This function is nxsched_running_setpriority, and the parameter tcb is always the running task. In this branch cpu == this_cpu(), so tcb is the same as this_task().

hujun260 replied:
I would like to clarify something
old_tcb = this_task()
if(nxsched_reprioritize_rtr())
{
new_tcb = this_task()

now  old_tcb != new_tcb
}

jlaitine replied:
Correct; it was already wrong originally and I should have fixed it at the same. I'll come back to this PR during next week!

hujun260 commented on sched/sched/sched_setpriority.c:
```diff
@@ -231,8 +223,7 @@ static inline void nxsched_running_setpriority(FAR struct tcb_s *tcb,
 #ifdef CONFIG_SMP
           int cpu = this_cpu();
-          if (tcb->cpu != cpu &&
-              tcb->task_state == TSTATE_TASK_RUNNING)
```
why  remove  "tcb->task_state == TSTATE_TASK_RUNNING"

jlaitine replied:
It is always true, there is no point in checking again. We are in "running_setpriority". See where this is called from (in critical section).

hujun260 replied:
ok

hujun260 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -196,7 +196,7 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
    */
 
-  if (nxsched_islocked_tcb(this_task()))
+  if (nxsched_islocked_tcb(rtcb))
```
Can you remove g_pendingtasks In SMP ?

xiaoxiang781216 replied:
@hujun260 let's create a new patch for this.

jlaitine replied:
@xiaoxiang781216 , I marked this as draft last week and just today started working on the removal of g_pendingtasks for SMP (as I promised below). Please let me know if you are taking this over, so I will just close this PR!

xiaoxiang781216 replied:
nif you are still working on this, please go ahead, thanks.

fdcavalcanti commented:
Will test on ESP32, its been failing for some time.

jlaitine commented:
Now there is an initial version, with removed g_pendingtasks for SMP.

I didn't test it fully yet, except quick trials with ostest on rv-virt:smp64 and with real application on mpfs w. 4 cores (CONFIG_BUILD_FLAT and CONFIG_BUILD_KERNEL).

I will do more testing tomorrow.

hujun260 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,7 +114,87 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
+          nxsched_add_prioritized(rtcb, list_readytorun());
+
+          /* We should now have only the idle task assigned */
+
```
I don't think it's necessary to remove "rtcb" from "assigntasklist" and move it to "readytorun", as this might cause "rtcb" to move back and forth between different lists, which would consume performance.

If we have to do this, I think "assigntasklist" could to be a single node (store running tcb) instead of being a list.

jlaitine replied:
This crossed my mind as well.

I'll rather try to find a way where a task is not assigned to a CPU running a sched locked task; even though the sched lock is a transitional state, it would be ok to check it for other cpus as well during scheduling (in critical section), since when a task exists sched lock, it will try to reschedule anyhow.

Leaving a task in assigned list will cause a kind of priority inheritance - if the sched lock lasts long (this happens for example in process startups in memory protected builds), the task left in the assigned list would be "stuck" (as long as the other task keeps the sched lock) and not eglible for scheduling for other CPUs.

You are right that there is no need for having a list for assigned tasks (i.e. running tasks). I just left it there. It could be removed as well, if continuing on the path described above.

hujun260 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,7 +114,87 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
+   */
+
+  if (nxsched_islocked_tcb(rtcb) ||
+      btcb->sched_priority < rtcb->sched_priority)
```
btcb->sched_priority <= rtcb->sched_priority

hujun260 commented on sched/sched/sched_setpriority.c:
```diff
@@ -191,10 +133,11 @@ static inline void nxsched_running_setpriority(FAR struct tcb_s *tcb,
 
-      if (nxsched_islocked_tcb(rtcb))
+#ifndef CONFIG_SMP
+      if (nxsched_islocked_tcb(tcb))
```
What if it is SMP and the current core is sched_locked?

hujun260 commented on sched/sched/sched_suspend.c:
```diff
@@ -172,7 +173,11 @@ void nxsched_suspend(FAR struct tcb_s *tcb)
+          if (!dq_empty(list_readytorun()))
+#else
+          if (!dq_empty(list_pendingtasks()))
+#endif
```
I think should like this

#ifdef CONFIG_SMP
          if (!dq_empty(list_readytorun()) && !nxsched_islocked_tcb(rtcb))
#else
          if (!dq_empty(list_pendingtasks()) && !nxsched_islocked_tcb(rtcb))
#endif

hujun260 replied:
I think should like this

#ifdef CONFIG_SMP
          if (!nxsched_islocked_tcb(rtcb) || switch_needed)
#else
          if (!nxsched_islocked_tcb(rtcb) || switch_needed)
#endif

jlaitine commented:
@hujun260 ; I changed the erroneous places which you pointed out,

I also added one patch on the top, which tries to address your concern of moving tasks back-and-forth when sched-locked task is running on one tcb. Could you please check if that makes sense? My intention is to prevent tasks repeatedly being assigned to a CPU, which is running a locked task. I think the most appropriate way to deal with that is to treat a locked task as if it had the highest possible priority; In other words, don't select that CPU for other tasks. When the task exists schedlock, it will anyhow re-assign the tasks to the CPUs.

As you suggested, the vector of assigned task lists could be changed into simple vector of  "g_running_task[CONFIG_SMP_NCPUS]" or such, because now the assigned tasks list only keeps the idle task in the end and one runnning task in front of that. I'd like to leave that to another PR, as it is only a cleanup and wouldn't change the functionality.

hujun260 commented:
> @hujun260 ; I changed the erroneous places which you pointed out,
> 
> I also added one patch on the top, which tries to address your concern of moving tasks back-and-forth when sched-locked task is running on one tcb. Could you please check if that makes sense? My intention is to prevent tasks repeatedly being assigned to a CPU, which is running a locked task. I think the most appropriate way to deal with that is to treat a locked task as if it had the highest possible priority; In other words, don't select that CPU for other tasks. When the task exits schedlock, it will anyhow re-assign the tasks to the CPUs.
> 
> As you suggested, the vector of assigned task lists could be changed into simple vector of "g_running_task[CONFIG_SMP_NCPUS]" or such, because now the assigned tasks list only keeps the idle task in the end and one runnning task in front of that. I'd like to leave that to another PR, as it is only a cleanup and wouldn't change the functionality.

From my experience, nxsched_add_prioritized is quite time-consuming. Several optimizations have already been made in this area specifically to reduce the number of calls to nxsched_add_prioritized. However, your changes have increased performance overhead. Especially when there are many tasks, during task switching, nxsched_add_prioritized is now called almost every time compared to before. Therefore, I hope you can submit the subsequent optimization code together so that we can conduct an overall evaluation.

jlaitine commented:
I realized that nxsched_remove_running had a lot of dead/unnecessary code now. Removed that...

jlaitine commented on sched/sched/sched_setpriority.c:
```diff
@@ -223,29 +230,30 @@ static inline void nxsched_running_setpriority(FAR struct tcb_s *tcb,
               tcb->task_state == TSTATE_TASK_RUNNING)
             {
-              struct reprioritize_arg_s arg;
+              struct reprioritize_arg_s *arg = &g_reprioritize_arg[cpu];
```
It is the same situation if you sent many separate priority change requests in a row synchronously; the last one will affect.

xiaoxiang781216 replied:
but the different target thread shouldn't lose the value.

jlaitine replied:
@xiaoxiang781216 But I now realized that the original code was correct. I just forgot how I thought it over.

The smp call is only sent to the one and only running task on the CPU, and the task is locked to that CPU before sending the message. The subsequent calls just replace the message, but they are all for the same tcb.

So it is not possible that the priority change messages are sent to other TCBs before the one is handled.

Maybe I should change this back, and not start adding stuff to the TCB...

xiaoxiang781216 replied:
I can't find a better solution too.

hujun260 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,14 +114,95 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
+   * running) and return false
+   */
+
+  if (nxsched_islocked_tcb(rtcb) ||
+      btcb->sched_priority <= rtcb->sched_priority ||
+      ((1 << cpu) & btcb->affinity) == 0)
```
These judgments are not necessary if called from nxsched_remove_self or sched_unlock.

jlaitine commented:
I am converting this back to draft, I don't have currently time to continue on this.

Will come back to this later

jlaitine commented:
I got this idea of changing the way how tasks are assigned to the CPUs.

I don't believe there is any reason to try to force a specific tcb for another CPU; rather we could just keep all the pending tasks in ready-to-run list, and let each CPU pick one to run from the top themselves.

Then we could change also all the priority adjustment, round robin etc. accross CPUs to just use the same smp_sched call.

This change would be rather large though. I just made it; keeping it as draft for now as I have limited time to work on this. Initially it feels stable and is performant, but I'll need to do more testing on other targets still.

jlaitine commented:
I guess it is ready for review. Re-tested on mpfs and risc-v rv-virt64.

hujun260 commented:
> I got this idea of changing the way how tasks are assigned to the CPUs.
> 
> I don't believe there is any reason to try to force a specific tcb for another CPU; rather we could just keep all the pending tasks in ready-to-run list, and let each CPU pick one to run from the top themselves.
> 
> Then we could change also all the priority adjustment, round robin etc. accross CPUs to just use the same smp_sched call.
> 
> This change would be rather large though. I just made it; keeping it as draft for now as I have limited time to work on this. Initially it feels stable and is performant, but I'll need to do more testing on other targets still.

Your modification seems too complicated and is likely to affect performance. Could you provide a comparison of the context switching performance data before and after the modification? Moreover, is it necessary to incorporate the Round Robin (RR) logic into addreadytorun as well?

hujun260 commented:
> Doesn't this actually simplify it a lot? It is basically removing a lot of redundant paths for scheduling between smp cores.
> 
> I only measured the overall performance on a system with several tasks processing data with high frequency, and see no change.
> 
> Do you have some specific tool in mind, to measure the context switching performance?
> 
> I'll try to make this smaller by keeping rr out of it, sometime next week!

Take the frequently called nxsched_add_readytorun/nxsched_process_delivered as examples.

Before your modification:

If a context switch occurs on the current CPU, there is no need to call any loop function.
If a context switch occurs on another CPU, the loop operation only needs to be called once.

After your modification:
At least two or three loop functions need to be called.

As for performance evaluation, it is actually not too difficult.

For example, there are 2 tasks, in two cases: on the same core or on different cores.

For instance, taskB is "waiting", and taskA initiates a "post" operation.

We need to calculate the time from when taskA initiates the operation to when taskB is awakened.

You can refer to the screenshots below
<img width="1203" height="638" alt="image" src="https://github.com/user-attachments/assets/d3da0394-49fc-440f-8002-7568fac67dd2" />

hujun260 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,14 +114,98 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
+   */
+
+  for (btcb = (FAR struct tcb_s *)dq_peek(list_readytorun());
+       btcb && btcb->sched_priority >= sched_priority;
```
btcb->sched_priority > sched_priority;

jlaitine replied:
The >= is on purpose. The switch_running is also used from roundrobin to avoid duplicating all this code in there. The intent for the function is to switch the running task on current CPU always when the priority is >=.

The priority is additionally checked before this function is called in places where equal priority task is not wanted to be switched out, for optimization.

jlaitine commented on sched/sched/sched_unlock.c:
```diff
@@ -69,10 +69,12 @@ void sched_unlock(void)
        * then pre-emption has been re-enabled.
        */
 
-      if (rtcb != NULL && --rtcb->lockcount == 0)
```
This task exists the sched lock at --rtcb->lockcount. Between that point, and the enter_critical_section below, there is a chance for an interrupt, which could trigger a re-schedule on this CPU,  leading this task to get scheduled out or migrate to another CPU

It is better to take the critical section already before setting lockcount to 0.

The issue doesn't occur when sched_unlock is called from semaphore code, since there we are already inside critical section. But from other places it happens.

hujun260 replied:
--rtcb->lockcount  is to decrement first, then reference.
if this task to get scheduled out，this task rtcb->lockcount must be zero

jlaitine replied:
1) Now you anyhow take the critical section, and typically take it without scheduling in between. There is no reason not to allow scheduling there, it is better to run it always the same way
2) In the next patch I am using the information of the rtcb->cpu, which might change if it schedules in between

So it is just cleaner to delay releasing the sched lock until inside the critical section. Then it works the same way in 100% of the cases instead of 99.999% of the cases 

Other than that, I am not sure if the current code would cause issues or not. Testing such thing is close to impossible, it just looks scary.

And yes, removing the critical section in the future is a good idea! ( and possibly removing the sched lock in semaphore code, would that be possible as well? )

jlaitine replied:
Here is a quick analysis of issues that the current code here causes:
1) This code would be in wrong place, causing something else accounted into being inside sched_lock:

```
/* Note that we no longer have pre-emption disabled. */

nxsched_critmon_preemption(rtcb, false, return_address(0));
sched_note_preemption(rtcb, false);

```

Later on there is the logic to check if there has been tasks pending because of the sched lock. Before this PR, there is a call to nxsched_merge_pending (and for single-cpu case there still is). This call is delayed if there is a schedule in between, leaving (possibly) higher priority task in pending list and scheduling something else in. So it is delaying the pending task schedule.

Something I still want to point out about the linux kernel comment; Linux scheduling supports many different scheduling schemes, and it is heavily optimized for CFQ (completely fair queuing). Many constructs are unnecessary and complicated if you just want to do priority based scheduling; this is a real-time os.

In addition, copying stuff from linux kernel is dangerous from the license point of view - linux is GPLv2. Knowing how the linux kernel works and using this knowledge to make own code is of course fine. But this is a fine line! Technically it is not even ok to study the linux kernel for the purpose of implementing the same in NuttX.

jlaitine commented:
I made some testing; executed some px4 based FC apps on MFPS with CONFIG_BUILD_KERNEL=y, CONFIG_SMP_NCPUS=4, CONFIG_DEBUG_ASSERTIONS=y, CONFIG_PRIORITY_INHERITANCE=y

This is a very semaphore heavy data-driven app, with data producers (sensors) frequencies varying between 800Hz to 2KHz (several pipelines running in parallel), and also communication between kernel and userspace is based on semaphores (+ lots of context switches).

With an earlier version of this PR (the SMP code is the same, just some cosmetic changes after that), the test has been succesfully running for 2 days 8 hours without errors:
```
nsh> uptime
15:25:05 up 2 days,  8:27
```

From the current PR version, I measured some CPU load with PX4 tools (px4 top):

px4 top:

```
 PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) TSLICE FD
   0 CPU0 IDLE                   77818 59.735   392/ 2048   0 (  0)      0  5
   1 CPU1 IDLE                   75126 74.300   632/ 2016   0 (  0)      0  5
   2 CPU2 IDLE                  104611 86.036   584/ 2016   0 (  0)      0  5
   3 CPU3 IDLE                  111112 92.352   584/ 2016   0 (  0)      0  5
...
Processes: 96 total, 5 running, 91 sleeping
CPU usage: 21.89% tasks, 0.00% sched, 78.11% idle
Uptime: 120.440s total, 77.818s idle
```

With the current upstream version of the scheduling, the same tool gives me 

```
 PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) TSLICE FD
   0 CPU0 IDLE                   86933 61.260   408/ 2048   0 (  0)      0  5
   1 CPU1 IDLE                  118468 73.182   616/ 2016   0 (  0)      0  5
   2 CPU2 IDLE                  134880 81.970   584/ 2016   0 (  0)      0  5
   3 CPU3 IDLE                  144453 91.641   584/ 2016   0 (  0)      0  5
...
Processes: 96 total, 5 running, 91 sleeping
CPU usage: 22.62% tasks, 0.00% sched, 77.38% idle
Uptime: 458.290s total, 276.382s idle
```

So initially it seems that the PR gives a small performance gain over the current version.

Additionally, I used the px4 "perf" tool, which measures a bunch of latencies and latency deviations, also these show some smaller deviations in latencies with this PR.

Also tested the rv-virt:smp64 target on qemu with ostest, which passes.
Additionally I enabled CONFIG_PRIORITY_INHERITANCE and CONFIG_DEBUG_ASSERTIONS on the rv-virt:smp64 target and checked with ostest.

fdcavalcanti commented:
> I made some testing; executed some px4 based FC apps on MFPS with CONFIG_BUILD_KERNEL=y, CONFIG_SMP_NCPUS=4, CONFIG_DEBUG_ASSERTIONS=y, CONFIG_PRIORITY_INHERITANCE=y
> 
> This is a very semaphore heavy data-driven app, with data producers (sensors) frequencies varying between 800Hz to 2KHz (several pipelines running in parallel), and also communication between kernel and userspace is based on semaphores (+ lots of context switches).
> 
> With an earlier version of this PR (the SMP code is the same, just some cosmetic changes after that), the test has been succesfully running for 2 days 8 hours without errors:
> 
> ```
> nsh> uptime
> 15:25:05 up 2 days,  8:27
> ```
> 
> From the current PR version, I measured some CPU load with PX4 tools (px4 top):
> 
> px4 top:
> 
> ```
>  PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) TSLICE FD
>    0 CPU0 IDLE                   77818 59.735   392/ 2048   0 (  0)      0  5
>    1 CPU1 IDLE                   75126 74.300   632/ 2016   0 (  0)      0  5
>    2 CPU2 IDLE                  104611 86.036   584/ 2016   0 (  0)      0  5
>    3 CPU3 IDLE                  111112 92.352   584/ 2016   0 (  0)      0  5
> ...
> Processes: 96 total, 5 running, 91 sleeping
> CPU usage: 21.89% tasks, 0.00% sched, 78.11% idle
> Uptime: 120.440s total, 77.818s idle
> ```
> 
> With the current upstream version of the scheduling, the same tool gives me
> 
> ```
>  PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) TSLICE FD
>    0 CPU0 IDLE                   86933 61.260   408/ 2048   0 (  0)      0  5
>    1 CPU1 IDLE                  118468 73.182   616/ 2016   0 (  0)      0  5
>    2 CPU2 IDLE                  134880 81.970   584/ 2016   0 (  0)      0  5
>    3 CPU3 IDLE                  144453 91.641   584/ 2016   0 (  0)      0  5
> ...
> Processes: 96 total, 5 running, 91 sleeping
> CPU usage: 22.62% tasks, 0.00% sched, 77.38% idle
> Uptime: 458.290s total, 276.382s idle
> ```
> 
> So initially it seems that the PR gives a small performance gain over the current version.
> 
> Additionally, I used the px4 "perf" tool, which measures a bunch of latencies and latency deviations, also these show some smaller deviations in latencies with this PR (something to note here: the "sched" figure doesn't give anything reasonable on risc-v due to placement of sched note hooks; all the time is accounted in the thread's execution times).
> 
> Also tested the rv-virt:smp64 target on qemu with ostest, which passes. Additionally I enabled CONFIG_PRIORITY_INHERITANCE and CONFIG_DEBUG_ASSERTIONS on the rv-virt:smp64 target and checked with ostest.
> 
> And finally; I am not experiencing the original issue ... which we first tried to patch, but then requests to remove the whole pending lists etc. lead to this re-write.
> 
> So all in all, I am currently quite happy with this...

Seems the tests have different uptimes. Can you repeat with a similar run time?
I'll test this on Espressif CI to so we have more answers on RISC-V and Xtensa.

hujun260 commented on sched/sched/sched_process_delivered.c:
```diff
@@ -78,67 +72,31 @@ void nxsched_process_delivered(int cpu)
-    }
-
-  if (nxsched_islocked_tcb(tcb))
+  if (!nxsched_switch_running(cpu))
```
Is it necessary to add a check here to ensure that tasks of the same priority are not replaced?
Although a check was performed before the interrupt was issued, once the critical section is exited, the state can change.

jlaitine replied:
I'll just add a parameter to nxsched_switch_running, to tell whether to compare >= or > "bool switch_equal", and set that in roundrobin. The code still needs to traverse the list in case affinity is restricted, so it is just easier/cleaner to be inside the nxsced_switch_running...

hujun260 commented on sched/sched/sched_unlock.c:
```diff
@@ -69,28 +76,41 @@ void sched_unlock(void)
+           */
+
+          ptcb = (FAR struct tcb_s *)dq_peek(list_readytorun());
+          if (ptcb && nxsched_select_cpu(ALL_CPUS) == rtcb->cpu &&
+              ptcb->sched_priority > rtcb->sched_priority &&
```
if (ptcb && ptcb->sched_priority > rtcb->sched_priority  &&
               nxsched_select_cpu(ALL_CPUS) == rtcb->cpu &&

jlaitine replied:
Good idea, will save a few cpu cycles also here

jlaitine commented:
Thanks @fdcavalcanti for testing! I'll check it tomorrow!

jlaitine commented on sched/sched/sched.h:
```diff
@@ -518,23 +514,51 @@ static inline_function bool nxsched_add_prioritized(FAR struct tcb_s *tcb,
+   * no need to do anything now.
+   */
+
+  if (!g_delivertasks[cpu])
```
I think this is necessary optimization. At least on my risc-v platform it even shows in the cpu load

xiaoxiang781216 replied:
should we change to:
```
if (cpu == curr_cpu)
  {
    ret = nxsched_switch_running(cpu);
  }
else if (!g_delivertasks[cpu])
  {
    g_delivertasks[cpu] = true;
    up_send_smp_sched(cpu);
  }
```

xiaoxiang781216 commented on sched/sched/sched_suspend.c:
```diff
@@ -172,9 +176,13 @@ void nxsched_suspend(FAR struct tcb_s *tcb)
           switch_needed = nxsched_remove_readytorun(tcb);
 
-          if (list_pendingtasks()->head)
+          if (!nxsched_islocked_tcb(rtcb) || switch_needed)
```
why not check switch_needed first

jlaitine replied:
sure can, this form was just due to an earlier review comment. I'll check switch needed first!

jlaitine commented on sched/sched/sched_unlock.c:
```diff
@@ -81,18 +89,28 @@ void sched_unlock(void)
+              nxsched_deliver_task(rtcb->cpu, rtcb->cpu))
+#else
+          ptcb = (FAR struct tcb_s *)dq_peek(list_pendingtasks());
+          if (ptcb && nxsched_merge_pending())
```
It just was like that originally, I tried not to change anything related single-cpu case in this PR.

jlaitine replied:
I'll just remove it then!

jlaitine commented:
@fdcavalcanti I tried to re-produce the issues, which you reported, in qemu.

I was not able to reproduce the DEBUGASSERT, which you reported, but I was able to get a double-interrupt crash in signest test.

There it seems as if something breaks the critical section; the issue seems to be fixed if I just add an extra critical section in/around the "nxsched_add_readytorun". But this doesn't make any sense, this function is only called with critical section already held. And there is nothing in the function itself which would break it.

Also, I didn't yet find any reason why this scheduling PR would make that happen. Anyhow it seems obvious that the critical section is not working properly here. This is tricky and needs to be debugged more!

hujun260 commented on arch/xtensa/src/common/xtensa_sigdeliver.c:
```diff
@@ -154,12 +154,11 @@ void xtensa_sig_deliver(void)
-  rtcb->irqcount++;
-  leave_critical_section((regs[REG_PS]));
-  rtcb->irqcount--;
+  leave_critical_section(up_irq_save());
```
This is just another way of implementation. Isn't the old one also fine?

jlaitine replied:
No, it doesn't do the same thing. Bumping the irqcount doesn't prevent the leave_critical_section from disabling the local interrupts. It simply sets the up_irq_restore(flags); in the end... And this is the only thing needed here, keep the irqs disabled for this cpu.

hujun260 replied:
This patch needs to be submitted independently. It is a long-existing bug and has nothing to do with the SMP scheduling modification.

jlaitine replied:
I submitted a separate PR on this:
https://github.com/apache/nuttx/pull/16804

If you just call " leave_critical_section((regs[REG_PS]))", the leave_critical_section will set the irqflags to regs[REG_PS], in the end of leave_critical_section, regardless of the rtcb->irqcount...

I just verified that it indeed does incorrectly enable the interrupts on the current CPU at that spot.

hujun260 replied:
ok！ You're right.

jlaitine commented:
Just found time to fix the esp32 issues which appeared on qemu. I also fixed one misplaced debugassert, and addressed the review comments.

I should have more time to work on this again next week, in case more issues are found. So putting it back to ready for review.

Thanks for the reviews so far, I do understand that this kind of PRs are terrible to review...

I now re-tested the functionality on esp32 qemu, and with MPFS.

fdcavalcanti commented:
@jlaitine I'll run CI on this again.

fdcavalcanti commented:
Getting some errors still.
One example from `esp32s3-devkit:wifi`:
```
2025-08-06 15:49:21 total segments stored 6
2025-08-06 15:49:21 AB[CPU0] dump_assert_info: Current Version: NuttX  10.4.0 c373b5e63a Aug  6 2025 13:01:43 xtensa
2025-08-06 15:49:21 [CPU0] dump_assert_info: Assertion failed (1 << cpu) & tcb->affinity: at file: chip/esp32s3_spiflash.c:942 task(CPU0): D process: Kernel 0x40379224
2025-08-06 15:49:21 [CPU0] up_dump_register:    PC: 420394a1    PS: 00060a22
2025-08-06 15:49:21 [CPU0] up_dump_register:    A0: 80379c04    A1: 3fcae340    A2: 00000000    A3: 3fc9b6f0
2025-08-06 15:49:21 [CPU0] up_dump_register:    A4: 00000000    A5: 3fcae0b0    A6: 00000000    A7: 3fc9a4c8
2025-08-06 15:49:21 [CPU0] up_dump_register:    A8: 00000000    A9: 00060a22   A10: 00000000   A11: 00000000
2025-08-06 15:49:21 [CPU0] up_dump_register:   A12: 3fc92798   A13: 3fc91938   A14: 3fc9a4c8   A15: 3fcae0e0
2025-08-06 15:49:21 [CPU0] up_dump_register:   SAR: 0000001f CAUSE: 00706f5f VADDR: 66337830
2025-08-06 15:49:21 [CPU0] up_dump_register:  LBEG: 00000000  LEND: 00000000  LCNT: 00000000
2025-08-06 15:49:21 [CPU0] dump_stacks: ERROR: Stack pointer 3fcae340is not within the stack
2025-08-06 15:49:21 [CPU0] dump_stackinfo: IRQ Stack:
2025-08-06 15:49:21 [CPU0] dump_stackinfo:   base: 0x40000100
2025-08-06 15:49:21 [CPU0] dump_stackinfo:   size: 00002048
```
And one from `esp32-devkitc:wifi_smp`:
```
2025-08-06 15:43:02 total segments stored 7
2025-08-06 15:43:02 A[    0.000000] [CPU0] __esp32_start: ESP32 chip revision is v3.0
2025-08-06 15:43:02 B[    0.010000] [CPU0] dump_assert_info: Current Version: NuttX  10.4.0 c373b5e63a Aug  6 2025 13:04:50 xtensa
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_assert_info: Assertion failed (1 << cpu) & tcb->affinity: at file: chip/esp32_spiflash.c:2473 task(CPU0): ??jA?0742??$pA? process: Kernel 0x40099090
2025-08-06 15:43:02 [    0.010000] [CPU0] up_dump_register:    PC: 40099bed    PS: 00060622
2025-08-06 15:43:02 [    0.010000] [CPU0] up_dump_register:    A0: 8009a161    A1: 3ffafdd0    A2: 00000000    A3: 3ffbade0
2025-08-06 15:43:02 [    0.010000] [CPU0] up_dump_register:    A4: 00000000    A5: 3ffafb40    A6: 00000000    A7: 3ffb97ec
2025-08-06 15:43:02 [    0.010000] [CPU0] up_dump_register:    A8: 00000000    A9: 00060622   A10: 00000000   A11: 00000000
2025-08-06 15:43:02 [    0.010000] [CPU0] up_dump_register:   A12: 3ffb17ac   A13: 3ffb0a38   A14: 3ffb97ec   A15: 3ffafb70
2025-08-06 15:43:02 [    0.010000] [CPU0] up_dump_register:   SAR: 0000001f CAUSE: deadbeef VADDR: deadbeef
2025-08-06 15:43:02 [    0.010000] [CPU0] up_dump_register:  LBEG: 00000000  LEND: 00000000  LCNT: 00000000
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stacks: ERROR: Stack pointer 3ffafdd0is not within the stack
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stackinfo: IRQ Stack:
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stackinfo:   base: 0x3ffb856c
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stackinfo:   size: 00002048
2025-08-06 15:43:02 [    0.010000] [CPU0] stack_dump: 0x3ffb8d44: 00018002 deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef 00000000
2025-08-06 15:43:02 [    0.010000] [CPU0] stack_dump: 0x3ffb8d64: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stacks: ERROR: Stack pointer 0 is not within the stack
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stackinfo: User Stack:
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stackinfo:   base: 0x31
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_stackinfo:   size: 00000008
2025-08-06 15:43:02 [    0.010000] [CPU0] stack_dump: 0x39: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2025-08-06 15:43:02 [    0.010000] [CPU0] dump_fatal_info: Dump CPU1: PAUSED
```

QEMU tests are looking ok now, but still, lots of crashes on different defconfigs (altough seems like it happens on the same place).
And that's just on Espressif devices. Maybe someone is testing on different archs?

jlaitine commented:
@fdcavalcanti : I made some code review around the related code paths. There seems to be several existing issues related to this, and I honestly don't understand how it has ever worked.

Observations from the sched/sched:

1) nxsched_set_affinity tries to schedule the task out and back in by calling nxsched_set_priority. But this works *only* if there happens to be another ready-to-run pending task with equal (or higher) priority in queue. And this is wrong. If the task is running on a wrong CPU, it should be scheduled out regardless of it's priority and then try to schedule it back (on another CPU). I have *not*  touched this behaviour... But for sure I can try to fix this. But this is really a different issue.

These are the observations from both the esp32_spiflash.c and the esp32s3_spiflash.c:

2) there is a race condition between the debugassert and nxsched_set_priority and esp32*spiflash.c. Since the DEBUGASSERT in esp32*spiflash.c is not inside a critical section, it is possible that the the "spiflash_op" thread hits the DEBUGASSERT (on one CPU) right after the the "nxsched_set_affinity" has set the new affinity mask, but before it has scheduled the other task out (on the other CPU).

3) in any case, the "spi_flash_op_block_task" may start on other CPU than what will be set later on in set_affinity. This means that the "int cpu = this_cpu();" just reads out the CPU on which the task is started, and later even if the task would have switched to the correct cpu (after the affinity setting), it is comparing just to the CPU on which it started on.

4) the DEBUGASSERT itself is crazy. It is only checking if the affinity mask of the tcb changes, and if the CPU in the affinity mask matches the CPU on which the task started on. It is not checking on which CPU the task is running on inside the loop, which I believe has been the intention! (it is not checking this_cpu() again inside the loop).

So, in order to fix this;

- In the esp32s3 you'd want to make sure that the "spi_flash_op_block_task" does not even start on a wrong CPU. You should be able to fix this by keeping a both "kthread_create" and "nxsched_set_affinity" inside a critical section in "spiflash_init_spi_flash_op_block_task". This should fix the obvious bug(s) in the spiflash driver.
- Perhaps think again what the DEBUGASSERT in question really wants to check?

In addition:

- The nxsched_set_affinity should be fixed as well; although this is an old bug, and doesn't cause this issue (it has been running on luck?)
- Are you sure that the CONFIG_ESP32_SPIFLASH_OP_TASK_STACKSIZE is large enough with debugasserts enabled? The stack dumps and especially the register values in the last dump look really suspicious.....

I am sorry I have caused extra testing effort for you, @fdcavalcanti; I really appreciate your testing! I definitely don't want to break things. It seems that I just managed to surface some hidden bugs with this PR... Like the signal delivery race condition which was really fatal and on many platforms, and we were really lucky that it started crashing on xtensa in qemu on that.

fdcavalcanti commented:
> @fdcavalcanti : I made some code review around the related code paths. There seems to be several existing issues related to this, and I honestly don't understand how it has ever worked.
> 
> Observations from the sched/sched:
> 
>     1. nxsched_set_affinity tries to schedule the task out and back in by calling nxsched_set_priority. But this works _only_ if there happens to be another ready-to-run pending task with equal (or higher) priority in queue. And this is wrong. If the task is running on a wrong CPU, it should be scheduled out regardless of it's priority and then try to schedule it back (on another CPU). I have _not_  touched this behaviour... But for sure I can try to fix this. But this is really a different issue.
> 
> 
> These are the observations from both the esp32_spiflash.c and the esp32s3_spiflash.c:
> 
>     2. there is a race condition between the debugassert and nxsched_set_priority and esp32_spiflash.c. Since the DEBUGASSERT in esp32_spiflash.c is not inside a critical section, it is possible that the the "spiflash_op" thread hits the DEBUGASSERT (on one CPU) right after the the "nxsched_set_affinity" has set the new affinity mask, but before it has scheduled the other task out (on the other CPU).
> 
>     3. in any case, the "spi_flash_op_block_task" may start on other CPU than what will be set later on in set_affinity. This means that the "int cpu = this_cpu();" just reads out the CPU on which the task is started, and later even if the task would have switched to the correct cpu (after the affinity setting), it is comparing just to the CPU on which it started on.
> 
>     4. the DEBUGASSERT itself is crazy. It is only checking if the affinity mask of the tcb changes, and if the CPU in the affinity mask matches the CPU on which the task started on. It is not checking on which CPU the task is running on inside the loop, which I believe has been the intention! (it is not checking this_cpu() again inside the loop).
> 
> 
> So, in order to fix this;
> 
>     * In the esp32s3 you'd want to make sure that the "spi_flash_op_block_task" does not even start on a wrong CPU. You should be able to fix this by keeping a both "kthread_create" and "nxsched_set_affinity" inside a single critical section in "spiflash_init_spi_flash_op_block_task". This should fix the obvious bug(s) in the spiflash driver.
> 
>     * Perhaps think again what the DEBUGASSERT in question really wants to check?
> 
> 
> In addition:
> 
>     * The nxsched_set_affinity should be fixed as well; although this is an old bug, and doesn't cause this issue (it has been running on luck?)
> 
>     * Are you sure that the CONFIG_ESP32_SPIFLASH_OP_TASK_STACKSIZE is large enough with debugasserts enabled? The stack dumps and especially the register values in the last dump look really suspicious.....
> 
> 
> I am sorry I have caused extra testing effort for you, @fdcavalcanti; I really appreciate your testing! I definitely don't want to break things. It seems that I just managed to surface some hidden bugs with this PR... Like the signal delivery race condition which was really fatal and on many platforms, and we were really lucky that it started crashing on xtensa in qemu on that.

Thank you for the detailed answer @jlaitine. Let us review this internally during the following week.
I would ask to not merge this PR for now.

jlaitine commented:
> > @fdcavalcanti : I made some code review around the related code paths. There seems to be several existing issues related to this, and I honestly don't understand how it has ever worked.
> > Observations from the sched/sched:
> > ```
> > 1. nxsched_set_affinity tries to schedule the task out and back in by calling nxsched_set_priority. But this works _only_ if there happens to be another ready-to-run pending task with equal (or higher) priority in queue. And this is wrong. If the task is running on a wrong CPU, it should be scheduled out regardless of it's priority and then try to schedule it back (on another CPU). I have _not_  touched this behaviour... But for sure I can try to fix this. But this is really a different issue.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > These are the observations from both the esp32_spiflash.c and the esp32s3_spiflash.c:
> > ```
> > 2. there is a race condition between the debugassert and nxsched_set_priority and esp32_spiflash.c. Since the DEBUGASSERT in esp32_spiflash.c is not inside a critical section, it is possible that the the "spiflash_op" thread hits the DEBUGASSERT (on one CPU) right after the the "nxsched_set_affinity" has set the new affinity mask, but before it has scheduled the other task out (on the other CPU).
> > 
> > 3. in any case, the "spi_flash_op_block_task" may start on other CPU than what will be set later on in set_affinity. This means that the "int cpu = this_cpu();" just reads out the CPU on which the task is started, and later even if the task would have switched to the correct cpu (after the affinity setting), it is comparing just to the CPU on which it started on.
> > 
> > 4. the DEBUGASSERT itself is crazy. It is only checking if the affinity mask of the tcb changes, and if the CPU in the affinity mask matches the CPU on which the task started on. It is not checking on which CPU the task is running on inside the loop, which I believe has been the intention! (it is not checking this_cpu() again inside the loop).
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So, in order to fix this;
> > ```
> > * In the esp32s3 you'd want to make sure that the "spi_flash_op_block_task" does not even start on a wrong CPU. You should be able to fix this by keeping a both "kthread_create" and "nxsched_set_affinity" inside a single critical section in "spiflash_init_spi_flash_op_block_task". This should fix the obvious bug(s) in the spiflash driver.
> > 
> > * Perhaps think again what the DEBUGASSERT in question really wants to check?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In addition:
> > ```
> > * The nxsched_set_affinity should be fixed as well; although this is an old bug, and doesn't cause this issue (it has been running on luck?)
> > 
> > * Are you sure that the CONFIG_ESP32_SPIFLASH_OP_TASK_STACKSIZE is large enough with debugasserts enabled? The stack dumps and especially the register values in the last dump look really suspicious.....
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am sorry I have caused extra testing effort for you, @fdcavalcanti; I really appreciate your testing! I definitely don't want to break things. It seems that I just managed to surface some hidden bugs with this PR... Like the signal delivery race condition which was really fatal and on many platforms, and we were really lucky that it started crashing on xtensa in qemu on that.
> 
> Thank you for the detailed answer @jlaitine. Let us review this internally during the following week. I would ask to not merge this PR for now.

I agree, there is no hurry to merge this PR. I have two crashing bugs which this PR solves, but of course we need to be certain that we don't break things.

jlaitine commented:
I just realized that there was a (incredibly silly) bug in nxsched_process_delivered, which caused extra interrupts on those platforms which call that on every syscall or interrupt. Fixed that one.

xiaoxiang781216 commented:
@fdcavalcanti @jlaitine does xtensa crash fixed?

tmedicci reviewed (commented):
We found a new issue on ESP32-S3 with SMP + SPI Flash enabled:

The `spi_flash_op_block_task` doesn't run on the second core (it was created within the critical section). It keeps present on `g_readytorun` list (its affinity set to run only in the second core), but it's never scheduled to run on it. This also prevents other tasks to run (as this is the task with higher priority).

This can be reproduced by:
```
make -j distclean && ./tools/configure.sh esp32s3-devkit:nsh && kconfig-tweak -e ESP32S3_SPIFLASH && kconfig-tweak -e SMP && kconfig-tweak --set-val SMP_NCPUS 2 && make olddefconfig && make flash ESPTOOL_PORT=/dev/ttyUSB0 -s -j$(nproc)
```

I'm afraid that there are missing logic at `sched/sched/sched_unlock.c`, because if we replace the outer `sched_lock/unlock` to a critical section (removing it from the inner loop that creates these threads, it works as expected)

jlaitine commented:
@tmedicci , @fdcavalcanti , I believe the issue which broke the esp32 spiflash driver, which uses fixed affinity for some of the threads, should now be fixed.

In qemu the tasks start properly now even without modifications to the spiflash driver; however, my observation about starting the task and setting the affinity is still valid. There is a possibility for the task to start on a wrong CPU before the affinity gets set (because the kthread_create also starts the task). This only depends on the timings, how fast the task spawns on the other CPU before the affinity gets set. This can be overcome by the critical_section or by creating the task and setting affinity before activating it ( I quickly tried this, which does that: https://github.com/tiiuae/nuttx/commit/b4f195b1a7d7d7a2baed597d446e33ac9860db9e )

tmedicci commented:
@jlaitine Just tested it and we still have the same issue. Let me try to explain it further.

Currently, we are testing [this](https://github.com/tmedicci/nuttx/tree/fix_smp_scheduling) branch:  It's basically your branch + the suggested change about the spiflash driver tasks (https://github.com/tmedicci/nuttx/commit/95150eae198742196a63b04c18a5ffd3c204668f). It still fails to run the same example I reported yesterday (here: https://github.com/apache/nuttx/pull/16673#pullrequestreview-3117567657). The device initializes, but the `nsh>` freezes.

By checking at the debugger, I could see that only one of the `spi_flash_op_block_task` ran in the first core. The same branch didn't run at all in the second core. Interestingly, by halting the device after typing anything, I can check the following values for `g_readytorun` and `g_assignedtasks`:
```
+p g_readytorun
$1 = {head = 0x3fc92578, tail = 0x3fc911a8}
+p g_assignedtasks
$2 = {0x3fc8ae00 <g_idletcb>, 0x3fc8aed4 <g_idletcb+212>}
```

Please note that both cores are running the idle thread (`g_assignedtasks`), but we have two pending tasks at `g_readytorun` and, curiously, the head TCB's (`0x3fc92578`) refers to the `spi_flash_op_block_task` task that should have run in the second core. The other task at tail (`0x3fc911a8`) refers to the UART handler that should process the typed character and echoes it back to nsh.

> I didn't quite understand what you mean with inner/outer loops etc; so maybe share the modified spiflash driver code to help me understand the issue better?

I referred to the loop that creates both `spi_flash_op_block_task` [here](https://github.com/tmedicci/incubator-nuttx/blob/95150eae198742196a63b04c18a5ffd3c204668f/arch/xtensa/src/esp32s3/esp32s3_spiflash.c#L1465). Please check that the scheduler is locked before the loop starts (and unlocked when it finishes). Also, please note that - as suggested - we protected the thread's creation/affinity within the critical section at [`spiflash_init_spi_flash_op_block_task`](https://github.com/tmedicci/incubator-nuttx/blob/95150eae198742196a63b04c18a5ffd3c204668f/arch/xtensa/src/esp32s3/esp32s3_spiflash.c#L1013).

In order to test it, I applied the following patch:
```
diff --git a/arch/xtensa/src/esp32s3/esp32s3_spiflash.c b/arch/xtensa/src/esp32s3/esp32s3_spiflash.c
index b34cead684..f83066674a 100644
--- a/arch/xtensa/src/esp32s3/esp32s3_spiflash.c
+++ b/arch/xtensa/src/esp32s3/esp32s3_spiflash.c
@@ -1004,14 +1004,11 @@ static int spiflash_init_spi_flash_op_block_task(int cpu)
   char *argv[2];
   char arg1[32];
   cpu_set_t cpuset;
-  irqstate_t flags;
 
   snprintf(arg1, sizeof(arg1), "%p", &cpu);
   argv[0] = arg1;
   argv[1] = NULL;
 
-  flags = enter_critical_section();
-
   pid = kthread_create("spiflash_op",
                        SCHED_PRIORITY_MAX,
                        CONFIG_ESP32S3_SPIFLASH_OP_TASK_STACKSIZE,
@@ -1031,8 +1028,6 @@ static int spiflash_init_spi_flash_op_block_task(int cpu)
       ret = -EPERM;
     }
 
-  leave_critical_section(flags);
-
   return ret;
 }
 #endif /* CONFIG_SMP */
@@ -1460,7 +1455,7 @@ int esp32s3_spiflash_init(void)
   nxrmutex_init(&g_flash_op_mutex);
 
 #ifdef CONFIG_SMP
-  sched_lock();
+  irqstate_t flags = enter_critical_section();
 
   for (cpu = 0; cpu < CONFIG_SMP_NCPUS; cpu++)
     {
@@ -1473,7 +1468,7 @@ int esp32s3_spiflash_init(void)
         }
     }
 
-  sched_unlock();
+  leave_critical_section(flags);
 #else
   UNUSED(cpu);
 #endif
```

Just removed the critical section from `spiflash_init_spi_flash_op_block_task` and protected the whole `spiflash_init_spi_flash_op_block_task` block outside the loop, by replacing the `sched_lock/unlock` for the critical section. **That worked!**

So, in summary, although this last patch made it work, it still seems to have something wrong with `sched_lock/unlock` because - without this last patch - the `spi_flash_op_block_task` for the second core remains forever on `g_readytorun` and prevents other tasks on that list from running too...

jlaitine commented:
@tmedicci ; would you be able to test it with the one patch I pushed on top of this branch? I still need to test and review it more thoroughly so it is not ready to merge as is. I would like to hear how it works on real esp32 HW.

I just tested it on some other HW's that it shouldn't break anything.

tmedicci commented:
> @tmedicci ; would you be able to test it with the one patch I pushed on top of this branch? I still need to test and review it more thoroughly so it is not ready to merge as is. I would like to hear how it works on real esp32 HW.
> 
> I just tested it on some other HW's that it shouldn't break anything.

I can confirm that the test described at https://github.com/apache/nuttx/pull/16673#issuecomment-3188301312 **now runs successfully** (after applying https://github.com/apache/nuttx/pull/16673/commits/bd71979987f2717e6354e3c8cea2a6fc09d461bf) even without my diff patch. I just triggered a new pipeline to evaluate the overall performance.

tmedicci commented:
> > @tmedicci ; would you be able to test it with the one patch I pushed on top of this branch? I still need to test and review it more thoroughly so it is not ready to merge as is. I would like to hear how it works on real esp32 HW.
> > I just tested it on some other HW's that it shouldn't break anything.
> 
> I can confirm that the test described at [#16673 (comment)](https://github.com/apache/nuttx/pull/16673#issuecomment-3188301312) **now runs successfully** (after applying [bd71979](https://github.com/apache/nuttx/commit/bd71979987f2717e6354e3c8cea2a6fc09d461bf)) even without my diff patch. I just triggered a new pipeline to evaluate the overall performance.

Everything seems fine with the new approach. All tests have passed! @fdcavalcanti , can you please double check?

The branch used for that is https://github.com/tmedicci/nuttx/tree/fix_smp_scheduling_2

jlaitine commented:
> Latest branch tested by @tmedicci looks good to me, no issues noticed. Thanks @jlaitine.

Thank *you* @fdcavalcanti  and @tmedicci for the great help with testing! This rabbithole has been a bit deeper than I expected....

shtirlic commented:
JFYI @jlaitine https://lore.kernel.org/lkml/20250528080924.2273858-1-mingo@kernel.org/
```
This series implements an idea Peter Zijlstra has mentioned
before: standardize the Linux scheduler on the SMP variant to
drastically reduce the number of CONFIG_SMP blocks of code
and related complexity in the scheduler.
```

xiaoxiang781216 commented on sched/sched/sched_removereadytorun.c:
```diff
@@ -174,145 +175,44 @@ void nxsched_remove_running(FAR struct tcb_s *tcb)
-               * tasks, we can use the dq_rem_mid macro to delete it.
-               */
-
-              dq_rem_mid(rtrtcb);
```
let's create a new patch remove dq_rem_head, dq_rem_mid and dq_addfirst_nonempty

xiaoxiang781216 commented on sched/sched/sched_process_delivered.c:
```diff
@@ -62,83 +62,46 @@
   DEBUGASSERT(g_cpu_nestcount[cpu] == 0);
   DEBUGASSERT(up_interrupt_context());
 
+  enum task_deliver_e priority;
```
move before line 65

xiaoxiang781216 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,170 +114,167 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
  * Returned Value:
- *   true if the currently active task (the head of the ready-to-run list)
- *   has changed.
+ *   true if the currently active task is switched to the btcb
```
but no btcb anymore

xiaoxiang781216 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,170 +114,167 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
+      if (tcb->sched_priority < btcb->sched_priority)
+        {
+          doswitch = nxsched_deliver_task(this_cpu(), target_cpu,
+                                              SWITCH_HIGHER);
```
align

hujun260 commented on sched/sched/sched.h:
```diff
@@ -549,7 +586,9 @@ static inline_function int nxsched_select_cpu(cpu_set_t affinity)
-          else if (rtcb->sched_priority <= minprio)
+          else if (rtcb->sched_priority <= minprio &&
+                   !nxsched_islocked_tcb(rtcb) &&
+                   (rtcb->flags & TCB_FLAG_CPU_LOCKED) == 0)
```
remove (rtcb->flags & TCB_FLAG_CPU_LOCKED) == 0

jlaitine replied:
The same as above, currently if the task is running, it can't be scheduled out, so this is correct...

jlaitine commented:
No change, no issues found. Re-based on top of master to see if the CI now passes.

jlaitine commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,170 +114,157 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
-
-  if (rtcb->sched_priority < btcb->sched_priority)
+  if (nxsched_islocked_tcb(rtcb) ||
+      (!is_idle_task(rtcb) && (rtcb->flags & TCB_FLAG_CPU_LOCKED) != 0))
```
that's not how it is used, for example in signal handling. There it specifically means that if the task is running, it must be also running when the SMP interrupt is executed (also now, before this change). Making the change you suggest would break at least the signal delivery.

This flag is only used when there is an SMP interrupt scheduled immediately, so it is fine to use it like that.

But I see that there is (and has been) a bit of a mixup with this flag, actually the whole flag might be redundant. IMHO, you'd only need affinity and lockcount.

Anyhow, I don't want to change this existing functionality for now...

jlaitine replied:
Right, I will re-check this part, thanks for the comment!

jlaitine commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,170 +114,160 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
-      btcb->task_state = TSTATE_TASK_RUNNING;
+  nxsched_add_prioritized(btcb, list_readytorun());
+
+  int target_cpu = btcb->flags & TCB_FLAG_CPU_LOCKED ? btcb->cpu :
```
I'd still disagree on that; if the TCB_FLAG_CPU_LOCKED acts as said in comments/documentation, it is supposed to override the affinity bits. That is, even if you didn't set the affinity, the CPU_LOCKED bit should work?

See e.g.
https://github.com/apache/nuttx/blob/e953bbd47c36d4b3cb5a38141de2595ffdcd467b/sched/init/nx_start.c#L395

hujun260 replied:
ok  
Encapsulate tcb->flags & TCB_FLAG_CPU_LOCKED into a macro and remove unnecessary affinity modifications.
<img width="894" height="444" alt="image" src="https://github.com/user-attachments/assets/8158adf5-9c7a-4204-9309-2e0e3629c4dc" />

jlaitine replied:
I'll remove the unnecessary affinity toggling. I don't think the check needs to be encapsulated into a macro; no other TCB flag checkings are done with macros.

Changing the check into a macro would mean that you'd need to also change setting and resetting the bit also with macros. And all of this would just hide it from the reader that you are manipulating a single bit in tcb flags.

jlaitine replied:
@hujun260 I added a cleanup patch on top, I believe this is what you ment?

hujun260 replied:
ok

xiaoxiang781216 commented on sched/sched/sched_addreadytorun.c:
```diff
@@ -114,170 +114,170 @@ bool nxsched_add_readytorun(FAR struct tcb_s *btcb)
-      DEBUGASSERT(task_state == TSTATE_TASK_RUNNING);
-      btcb->cpu        = cpu;
-      btcb->task_state = TSTATE_TASK_RUNNING;
+  int target_cpu = btcb->flags & TCB_FLAG_CPU_LOCKED ? btcb->cpu :
```
move before line 262 to avoid the variable definition in the middle of function

jlaitine replied:
Yes, thanks. I also improved the comment a bit

jlaitine commented:
Big thanks to everyone who helped in testing and to the reviewers!

fdcavalcanti commented:
> > @jlaitine hi! I noticed this is again breaking the same configs on ESP devices.
> > Seems some changes were made after we tested and it broke again.
> > Please let's take a look at it, otherwise my suggestion is to revert those changes.
> > ```
> > 2025-08-21 19:39:36 [CPU0] dump_assert_info: Assertion failed (1 << cpu) & tcb->affinity: at file: chip/esp32s3_spiflash.c:942
> > ```
> 
> Yes, let's just fix it. This is the assertion, which I mentioned before that is inherently broken, and passes just by luck. If the thread happens to reach this point after creation, before the affinity gets set, it will assert here.
> 
> One trick would be to move the assertion (_and_ this_cpu()) inside the while loop after the semaphore wait. Other, more proper, solutions would be 1) create the thread, set the affinity and then activate it. I have a patch for this ready. 2) Enclose creation and affinity setting in a critical section.
> 
> We could both move the debugassert (so that it actually makes sense....) and fix the task creation/affinity set sequence.
> 
> I will of course analyze the review changes made after testing it!
> 
> Edit: These were the existing suggested solutions, which both fix the existing bug in the spiflash_driver:
> 
> #16894 #16895
> 
> One is your patch, @fdcavalcanti. My solution is a bit longer but doesn't rely on the critical section, only does things in the proper order (like e.g. pthread_create does, just that there is no such interface for kthread_create).
> 
> I didn't see anything immediately wrong in the last changes done in this PR. On the contrary, they look correct. But could actually cause a thread to start faster after creation. But of course, if the above solutions don't work then there must be something wrong still introduced in this PR

Thank you for this @jlaitine, I'll test both PRs internally and answer on the proper page!
